### PR TITLE
Added inline type annotations to Babel

### PR DESCRIPTION
**Pull request by DenverCoder1**
This PR adds inline type annotations to the public methods and classes in Babel.

This PR tries to focus on only _adding_ correct type annotations. There are still some Pyright warnings due to non-type-safe implementations which can be fixed at a later time.

Most of the annotations have already been reviewed and approved by [typeshed](https://github.com/python/typeshed).

Tested on Python 3.7 - 3.11

### CODE REVIEW

File: babel/core.py [hunk 1]
```diff
@@ -217,8 +243,16 @@ def negotiate(cls, preferred, available, sep='_', aliases=LOCALE_ALIASES):
...
+    @classmethod
+    def parse(cls, identifier: None, sep: str = ..., resolve_likely_subtags: bool = ...) -> None: ...
+
+    @overload
+    @classmethod
+    def parse(cls, identifier: str | Locale, sep: str = ..., resolve_likely_subtags: bool = ...) -> Locale: ...
+
     @classmethod
-    def parse(cls, identifier, sep='_', resolve_likely_subtags=True):
+    def parse(cls, identifier: str | Locale | None, sep: str = '_', resolve_likely_subtags: bool = True) -> Locale | None:
```
[the remark this answers is not in the document]
Comment (akx) [reply babel/core.py:255]: As an aside, we should probably look at getting rid of `None` being an allowed `identifier` here.

File: babel/messages/extract.py [hunk 2]
```diff
@@ -41,38 +70,38 @@
...
-):
+    dirname: str | os.PathLike[str] | None = None,
+    method_map: Iterable[tuple[str, str]] = DEFAULT_MAPPING,
+    options_map: SupportsItems[str, dict[str, Any]] | None = None,
+    keywords: Mapping[str, _Keyword] = DEFAULT_KEYWORDS,
+    comment_tags: Collection[str] = (),
+    callback: Callable[[str, str, dict[str, Any]], object] | None = None,
+    strip_comment_tags: bool = False,
+    directory_filter: Callable[[str], bool] | None = None,
+) -> Generator[tuple[str, int, str | tuple[str, ...], list[str], str | None], None, None]:
```
Comment (akx) [inline babel/messages/extract.py:104]: Maybe this generator type should be type-aliased (maybe as `Iterable`?)

File: babel/messages/extract.py [hunk 3]
```diff
@@ -533,7 +585,9 @@ def _parse_python_string(value, encoding, future_flags):
     return None
 
 
-def extract_javascript(fileobj, keywords, comment_tags, options, lineno=1):
+def extract_javascript(
+    fileobj: _FileObj, keywords: Mapping[str, _Keyword], comment_tags: Collection[str], options: _JSOptions, lineno: int = 1
```
Comment (akx) [inline babel/messages/extract.py:589]:
```suggestion
    fileobj: _FileObj, 
    keywords: Mapping[str, _Keyword], 
    comment_tags: Collection[str], 
    options: _JSOptions, 
    lineno: int = 1,
```

File: babel/messages/mofile.py [hunk 4]
```diff
@@ -7,18 +7,22 @@
...
+from __future__ import annotations
 
 import array
 import struct
+from typing import TYPE_CHECKING
 
 from babel.messages.catalog import Catalog, Message
 
+if TYPE_CHECKING:
+    from _typeshed import SupportsRead, SupportsWrite
```
Comment (akx) [inline babel/messages/mofile.py:19]: Is this an okay import..? (Ditto for all other `_typeshed` imports)

File: babel/messages/pofile.py [hunk 5]
```diff
@@ -313,15 +322,17 @@ def parse(self, fileobj):
...
         if self.abort_invalid:
             raise PoFileError(msg, self.catalog, line, lineno)
         print("WARNING:", msg)
         print(f"WARNING: Problem on line {lineno + 1}: {line!r}")
 
 
-def read_po(fileobj, locale=None, domain=None, ignore_obsolete=False, charset=None, abort_invalid=False):
+def read_po(fileobj: Iterable[str | bytes], locale: str | Locale | None = None,
+            domain: str | None = None, ignore_obsolete: bool = False,
+            charset: str | None = None, abort_invalid: bool = False) -> Catalog:
```
Comment (akx) [inline babel/messages/pofile.py:335]:
```suggestion
def read_po(
    fileobj: Iterable[str | bytes],
    locale: str | Locale | None = None,
    domain: str | None = None,
    ignore_obsolete: bool = False,
    charset: str | None = None,
    abort_invalid: bool = False,
) -> Catalog:
```

File: babel/messages/pofile.py [hunk 6]
```diff
@@ -313,15 +322,17 @@ def parse(self, fileobj):
...
+    def _invalid_pofile(self, line, lineno, msg) -> None:
         assert isinstance(line, str)
         if self.abort_invalid:
             raise PoFileError(msg, self.catalog, line, lineno)
         print("WARNING:", msg)
         print(f"WARNING: Problem on line {lineno + 1}: {line!r}")
 
 
-def read_po(fileobj, locale=None, domain=None, ignore_obsolete=False, charset=None, abort_invalid=False):
+def read_po(fileobj: Iterable[str | bytes], locale: str | Locale | None = None,
```
Comment (akx) [inline babel/messages/pofile.py:333]: Is `Iterable[str | bytes]` actually correct here? It looks like it should be a filelike.

File: babel/messages/pofile.py [hunk 7]
```diff
@@ -460,9 +471,10 @@ def normalize(string, prefix='', width=76):
     return u'""\n' + u'\n'.join([(prefix + escape(line)) for line in lines])
 
 
-def write_po(fileobj, catalog, width=76, no_location=False, omit_header=False,
-             sort_output=False, sort_by_file=False, ignore_obsolete=False,
-             include_previous=False, include_lineno=True):
+def write_po(fileobj: SupportsWrite[bytes], catalog: Catalog, width: int = 76,
+             no_location: bool = False, omit_header: bool = False,
+             sort_output: bool = False, sort_by_file: bool = False, ignore_obsolete: bool = False,
+             include_previous: bool = False, include_lineno: bool = True) -> None:
```
Comment (akx) [inline babel/messages/pofile.py:477]:
```suggestion
def write_po(
    fileobj: SupportsWrite[bytes],
    catalog: Catalog,
    width: int = 76,
    no_location: bool = False,
    omit_header: bool = False,
    sort_output: bool = False,
    sort_by_file: bool = False,
    ignore_obsolete: bool = False,
    include_previous: bool = False,
    include_lineno: bool = True,
) -> None:
```

File: babel/numbers.py [hunk 8]
```diff
@@ -33,7 +38,7 @@ class UnknownCurrencyError(Exception):
     """Exception thrown when a currency is requested for which no data is available.
     """
 
-    def __init__(self, identifier):
+    def __init__(self, identifier: str):
```
Comment (akx) [inline babel/numbers.py:41]:
```suggestion
    def __init__(self, identifier: str) -> None:
```

File: babel/numbers.py [hunk 9]
```diff
@@ -102,7 +107,8 @@ def normalize_currency(currency, locale=None):
     return currency
 
 
-def get_currency_name(currency, count=None, locale=LC_NUMERIC):
+def get_currency_name(currency: str, count: float | decimal.Decimal | None = None,
+                      locale: Locale | str | None = LC_NUMERIC) -> str:
```
Comment (akx) [inline babel/numbers.py:111]:
```suggestion
def get_currency_name(
    currency: str,
    count: float | decimal.Decimal | None = None,
    locale: Locale | str | None = LC_NUMERIC,
) -> str:
```

File: babel/numbers.py [hunk 10]
```diff
@@ -154,7 +160,8 @@ def get_currency_precision(currency):
     return precisions.get(currency, precisions['DEFAULT'])[0]
 
 
-def get_currency_unit_pattern(currency, count=None, locale=LC_NUMERIC):
+def get_currency_unit_pattern(currency: str, count: float | decimal.Decimal | None = None,
+                              locale: Locale | str | None = LC_NUMERIC) -> str:
```
Comment (akx) [inline babel/numbers.py:164]:
```suggestion
def get_currency_unit_pattern(
    currency: str,
    count: float | decimal.Decimal | None = None,
    locale: Locale | str | None = LC_NUMERIC,
) -> str:
```

File: babel/numbers.py [hunk 11]
```diff
@@ -184,9 +191,9 @@ def get_currency_unit_pattern(currency, count=None, locale=LC_NUMERIC):
     return loc._data['currency_unit_patterns']['other']
 
 
-def get_territory_currencies(territory, start_date=None, end_date=None,
-                             tender=True, non_tender=False,
-                             include_details=False):
+def get_territory_currencies(territory: str, start_date: date_ | None = None, end_date: date_ | None = None,
+                             tender: bool = True, non_tender: bool = False,
+                             include_details: bool = False) -> list[str]:
```
Comment (akx) [inline babel/numbers.py:196]:
```suggestion
def get_territory_currencies(
    territory: str,
    start_date: date_ | None = None,
    end_date: date_ | None = None,
    tender: bool = True,
    non_tender: bool = False,
    include_details: bool = False,
) -> list[str]:
```

Comment (akx) [reply babel/numbers.py:196]: This doesn't seem to take into account `include_details` changing the rtype (which is dumb API design in the first place but here we are... :) )

File: babel/numbers.py [hunk 12]
```diff
@@ -363,14 +370,15 @@ def get_decimal_precision(number):
...
+def get_decimal_quantum(precision: int | decimal.Decimal) -> decimal.Decimal:
     """Return minimal quantum of a number, as defined by precision."""
     assert isinstance(precision, (int, decimal.Decimal))
     return decimal.Decimal(10) ** (-precision)
 
 
 def format_decimal(
-        number, format=None, locale=LC_NUMERIC, decimal_quantization=True, group_separator=True):
+        number: float | decimal.Decimal | str, format: str | None = None, locale: Locale | str | None = LC_NUMERIC,
+        decimal_quantization: bool = True, group_separator: bool = True) -> str:
```
Comment (akx) [inline babel/numbers.py:381]:
```suggestion
    number: float | decimal.Decimal | str,
    format: str | None = None,
    locale: Locale | str | None = LC_NUMERIC,
    decimal_quantization: bool = True,
    group_separator: bool = True,
) -> str:
```

File: babel/numbers.py [hunk 13]
```diff
@@ -419,7 +427,8 @@ def format_decimal(
         number, locale, decimal_quantization=decimal_quantization, group_separator=group_separator)
 
 
-def format_compact_decimal(number, *, format_type="short", locale=LC_NUMERIC, fraction_digits=0):
+def format_compact_decimal(number: float | decimal.Decimal | str, *, format_type: Literal['short', 'long'] = 'short',
+                           locale: Locale | str | None = LC_NUMERIC, fraction_digits: int = 0) -> str:
```
Comment (akx) [inline babel/numbers.py:431]:
```suggestion
def format_compact_decimal(
    number: float | decimal.Decimal | str,
    *,
    format_type: Literal["short", "long"] = "short",
    locale: Locale | str | None = LC_NUMERIC,
    fraction_digits: int = 0,
) -> str:
```

File: babel/numbers.py [hunk 14]
```diff
@@ -450,7 +459,8 @@ def format_compact_decimal(number, *, format_type="short", locale=LC_NUMERIC, fr
     return pattern.apply(number, locale, decimal_quantization=False)
 
 
-def _get_compact_format(number, compact_format, locale, fraction_digits=0):
+def _get_compact_format(number: float | decimal.Decimal | str, compact_format: LocaleDataDict,
+                        locale: Locale | str | None, fraction_digits: int) -> tuple[decimal.Decimal, NumberPattern | None]:
```
Comment (akx) [inline babel/numbers.py:463]:
```suggestion
def _get_compact_format(
    number: float | decimal.Decimal | str,
    compact_format: LocaleDataDict,
    locale: Locale | str | None,
    fraction_digits: int,
) -> tuple[decimal.Decimal, NumberPattern | None]:
```

File: babel/numbers.py [hunk 15]
```diff
@@ -488,8 +498,10 @@ class UnknownCurrencyFormatError(KeyError):
 
 
 def format_currency(
-        number, currency, format=None, locale=LC_NUMERIC, currency_digits=True,
-        format_type='standard', decimal_quantization=True, group_separator=True):
+        number: float | decimal.Decimal | str, currency: str, format: str | None = None,
+        locale: Locale | str | None = LC_NUMERIC, currency_digits: bool = True,
+        format_type: Literal['name', 'standard', 'accounting'] = 'standard',
+        decimal_quantization: bool = True, group_separator: bool = True) -> str:
```
Comment (akx) [inline babel/numbers.py:504]:
```suggestion
    number: float | decimal.Decimal | str,
    currency: str,
    format: str | None = None,
    locale: Locale | str | None = LC_NUMERIC,
    currency_digits: bool = True,
    format_type: Literal["name", "standard", "accounting"] = "standard",
    decimal_quantization: bool = True,
    group_separator: bool = True,
) -> str:
```

File: babel/numbers.py [hunk 16]
```diff
@@ -596,8 +608,10 @@ def format_currency(
 
 
 def _format_currency_long_name(
-        number, currency, format=None, locale=LC_NUMERIC, currency_digits=True,
-        format_type='standard', decimal_quantization=True, group_separator=True):
+        number: float | decimal.Decimal | str, currency: str, format: str | None = None,
+        locale: Locale | str | None = LC_NUMERIC, currency_digits: bool = True,
+        format_type: Literal['name', 'standard', 'accounting'] = 'standard',
+        decimal_quantization: bool = True, group_separator: bool = True) -> str:
```
Comment (akx) [inline babel/numbers.py:614]:
```suggestion
    number: float | decimal.Decimal | str,
    currency: str,
    format: str | None = None,
    locale: Locale | str | None = LC_NUMERIC,
    currency_digits: bool = True,
    format_type: Literal["name", "standard", "accounting"] = "standard",
    decimal_quantization: bool = True,
    group_separator: bool = True,
) -> str:
```

File: babel/numbers.py [hunk 17]
```diff
@@ -631,7 +645,9 @@ def _format_currency_long_name(
     return unit_pattern.format(number_part, display_name)
 
 
-def format_compact_currency(number, currency, *, format_type="short", locale=LC_NUMERIC, fraction_digits=0):
+def format_compact_currency(number: float | decimal.Decimal | str,
+        currency: str, *, format_type: Literal['short'] = 'short',
+        locale: Locale | str | None = LC_NUMERIC, fraction_digits: int = 0) -> str:
```
Comment (akx) [inline babel/numbers.py:650]:
```suggestion
def format_compact_currency(
    number: float | decimal.Decimal | str,
    currency: str,
    *,
    format_type: Literal["short"] = "short",
    locale: Locale | str | None = LC_NUMERIC,
    fraction_digits: int = 0
) -> str:
```

File: babel/numbers.py [hunk 18]
```diff
@@ -670,7 +686,9 @@ def format_compact_currency(number, currency, *, format_type="short", locale=LC_
 
 
 def format_percent(
-        number, format=None, locale=LC_NUMERIC, decimal_quantization=True, group_separator=True):
+        number: float | decimal.Decimal | str, format: str | None = None,
+        locale: Locale | str | None = LC_NUMERIC, decimal_quantization: bool = True,
+        group_separator: bool = True) -> str:
```
Comment (akx) [inline babel/numbers.py:691]:
```suggestion
    number: float | decimal.Decimal | str,
    format: str | None = None,
    locale: Locale | str | None = LC_NUMERIC,
    decimal_quantization: bool = True,
    group_separator: bool = True,
) -> str:
```

File: babel/numbers.py [hunk 19]
```diff
@@ -717,7 +735,8 @@ def format_percent(
 
 
 def format_scientific(
-        number, format=None, locale=LC_NUMERIC, decimal_quantization=True):
+        number: float | decimal.Decimal | str, format: str | None = None, locale: Locale | str | None = LC_NUMERIC,
+        decimal_quantization: bool = True) -> str:
```
Comment (akx) [inline babel/numbers.py:739]:
```suggestion
        number: float | decimal.Decimal | str,
        format: str | None = None,
        locale: Locale | str | None = LC_NUMERIC,
        decimal_quantization: bool = True,
) -> str:
```

File: babel/numbers.py [hunk 20]
```diff
@@ -754,13 +773,13 @@ def format_scientific(
 class NumberFormatError(ValueError):
     """Exception raised when a string cannot be parsed into a number."""
 
-    def __init__(self, message, suggestions=None):
+    def __init__(self, message: str, suggestions: str | None = None):
```
Comment (akx) [inline babel/numbers.py:776]:
```suggestion
    def __init__(self, message: str, suggestions: str | None = None) -> None:
```

File: babel/numbers.py [hunk 21]
```diff
@@ -970,9 +989,9 @@ def parse_precision(p):
 
 class NumberPattern:
 
-    def __init__(self, pattern, prefix, suffix, grouping,
-                 int_prec, frac_prec, exp_prec, exp_plus,
-                 number_pattern: str | None = None):
+    def __init__(self, pattern: str, prefix: tuple[str, str], suffix: tuple[str, str], grouping: tuple[int, int],
+                 int_prec: tuple[int, int], frac_prec: tuple[int, int], exp_prec: tuple[int, int] | None,
+                 exp_plus: bool | None, number_pattern: str | None = None):
```
Comment (akx) [inline babel/numbers.py:994]:
```suggestion
    def __init__(
        self,
        pattern: str,
        prefix: tuple[str, str],
        suffix: tuple[str, str],
        grouping: tuple[int, int],
        int_prec: tuple[int, int],
        frac_prec: tuple[int, int],
        exp_prec: tuple[int, int] | None,
        exp_plus: bool | None,
        number_pattern: str | None = None,
    ) -> None:
```

File: babel/plural.py [hunk 22]
```diff
@@ -97,18 +103,18 @@ class PluralRule:
 
     __slots__ = ('abstract', '_func')
 
-    def __init__(self, rules):
+    def __init__(self, rules: Mapping[str, str] | Iterable[tuple[str, str]]):
```
Comment (akx) [inline babel/plural.py:106]:
```suggestion
    def __init__(self, rules: Mapping[str, str] | Iterable[tuple[str, str]]) -> None:
```

File: babel/plural.py [hunk 23]
```diff
@@ -354,30 +363,29 @@ def tokenize_rule(s):
                             'Got unexpected %r' % s[pos])
     return result[::-1]
 
-
-def test_next_token(tokens, type_, value=None):
+def test_next_token(tokens: list[tuple[str, str]], type_: str, value: str | None = None) -> list[tuple[str, str]] | bool:
```
Comment (akx) [inline babel/plural.py:366]:
```suggestion
def test_next_token(
    tokens: list[tuple[str, str]],
    type_: str,
    value: str | None = None,
) -> list[tuple[str, str]] | bool:
```

File: babel/plural.py [hunk 24]
```diff
@@ -354,30 +363,29 @@ def tokenize_rule(s):
...
     return 'value', (value, )
 
 
-def ident_node(name):
+def ident_node(name: str) -> tuple[str, tuple[()]]:
     return name, ()
 
 
-def range_list_node(range_list):
+def range_list_node(range_list: Iterable[Iterable[float | decimal.Decimal]]) -> tuple[Literal['range_list'], Iterable[Iterable[float | decimal.Decimal]]]:
```
Comment (akx) [inline babel/plural.py:384]:
```suggestion
def range_list_node(
    range_list: Iterable[Iterable[float | decimal.Decimal]],
) -> tuple[Literal['range_list'], Iterable[Iterable[float | decimal.Decimal]]]:
```

File: babel/support.py [hunk 25]
```diff
@@ -34,7 +47,7 @@ class Format:
     u'1.234'
     """
 
-    def __init__(self, locale, tzinfo=None):
+    def __init__(self, locale: Locale | str, tzinfo: BaseTzInfo | None = None):
```
Comment (akx) [inline babel/support.py:50]:
```suggestion
    def __init__(self, locale: Locale | str, tzinfo: BaseTzInfo | None = None) -> None:
```

File: babel/support.py [hunk 26]
```diff
@@ -76,8 +89,10 @@ def time(self, time=None, format='medium'):
         """
         return format_time(time, format, tzinfo=self.tzinfo, locale=self.locale)
 
-    def timedelta(self, delta, granularity='second', threshold=.85,
-                  format='long', add_direction=False):
+    def timedelta(self, delta: _timedelta | int,
+                  granularity: Literal['year', 'month', 'week', 'day', 'hour', 'minute', 'second'] = 'second',
+                  threshold: float = .85, format: Literal['narrow', 'short', 'medium', 'long'] = 'long',
+                  add_direction: bool = False) -> str:
```
Comment (akx) [inline babel/support.py:95]:
```suggestion
    def timedelta(
        self,
        delta: _timedelta | int,
        granularity: Literal["year", "month", "week", "day", "hour", "minute", "second"] = "second",
        threshold: float = 0.85,
        format: Literal["narrow", "short", "medium", "long"] = "long",
        add_direction: bool = False,
    ) -> str:
```

File: babel/support.py [hunk 27]
```diff
@@ -108,7 +123,8 @@ def decimal(self, number, format=None):
         """
         return format_decimal(number, format, locale=self.locale)
 
-    def compact_decimal(self, number, format_type='short', fraction_digits=0):
+    def compact_decimal(self, number: float | decimal.Decimal | str,
+                        format_type: Literal['short', 'long'] = 'short', fraction_digits: int = 0) -> str:
```
Comment (akx) [inline babel/support.py:127]:
```suggestion
    def compact_decimal(
        self,
        number: float | decimal.Decimal | str,
        format_type: Literal['short', 'long'] = 'short',
        fraction_digits: int = 0,
    ) -> str:
```

File: babel/support.py [hunk 28]
```diff
@@ -119,19 +135,20 @@ def compact_decimal(self, number, format_type='short', fraction_digits=0):
...
 
-    def currency(self, number, currency):
+    def currency(self, number: float | decimal.Decimal | str, currency: str) -> str:
         """Return a number in the given currency formatted for the locale.
         """
         return format_currency(number, currency, locale=self.locale)
 
-    def compact_currency(self, number, currency, format_type='short', fraction_digits=0):
+    def compact_currency(self, number: float | decimal.Decimal | str, currency: str,
+                         format_type: Literal['short'] = 'short', fraction_digits: int = 0) -> str:
```
Comment (akx) [inline babel/support.py:144]:
```suggestion
    def compact_currency(
        self,
        number: float | decimal.Decimal | str,
        currency: str,
        format_type: Literal['short'] = 'short',
        fraction_digits: int = 0,
    ) -> str:
```

File: babel/support.py [hunk 29]
```diff
@@ -183,18 +200,25 @@ class LazyProxy:
...
-        is_cache_enabled = kwargs.pop('enable_cache', True)
+    if TYPE_CHECKING:
+        _func: Callable[..., Any]
+        _args: tuple[Any, ...]
+        _kwargs: dict[str, Any]
+        _is_cache_enabled: bool
+        _value: Any
+        _attribute_error: AttributeError | None
+
+    def __init__(self, func: Callable[..., Any], *args: Any, enable_cache: bool = True, **kwargs: Any):
```
Comment (akx) [inline babel/support.py:211]:
```suggestion
    def __init__(self, func: Callable[..., Any], *args: Any, enable_cache: bool = True, **kwargs: Any) -> None:
```

File: babel/support.py [hunk 30]
```diff
@@ -304,9 +328,13 @@ def __deepcopy__(self, memo):
...
 class NullTranslations(gettext.NullTranslations):
 
+    if TYPE_CHECKING:
+        _info: dict[str, str]
+        _fallback: NullTranslations | None
+
     DEFAULT_DOMAIN = None
 
-    def __init__(self, fp=None):
+    def __init__(self, fp: gettext._TranslationsReader | None = None):
```
Comment (akx) [inline babel/support.py:337]:
```suggestion
    def __init__(self, fp: gettext._TranslationsReader | None = None) -> None:
```

File: babel/support.py [hunk 31]
```diff
@@ -555,7 +584,9 @@ def __init__(self, fp=None, domain=None):
     ungettext = gettext.GNUTranslations.ngettext
 
     @classmethod
-    def load(cls, dirname=None, locales=None, domain=None):
+    def load(cls, dirname: str | os.PathLike[str] | None = None,
+             locales: list[str] | tuple[str, ...] | str | None = None,
+             domain: str | None = None) -> NullTranslations:
```
Comment (akx) [inline babel/support.py:589]:
```suggestion
    def load(
        cls,
        dirname: str | os.PathLike[str] | None = None,
        locales: list[str] | tuple[str, ...] | str | None = None,
        domain: str | None = None,
    ) -> NullTranslations:
```

File: babel/units.py [hunk 32]
```diff
@@ -1,13 +1,21 @@
...
 
 class UnknownUnitError(ValueError):
-    def __init__(self, unit, locale):
+    def __init__(self, unit: str, locale: Locale) -> None:
         ValueError.__init__(self, f"{unit} is not a known unit in {locale}")
 
 
-def get_unit_name(measurement_unit, length='long', locale=LC_NUMERIC):
+def get_unit_name(measurement_unit: str, length: Literal['short', 'long', 'narrow'] = 'long',
+                  locale: Locale | str | None = LC_NUMERIC) -> str | None:
```
Comment (akx) [inline babel/units.py:18]:
```suggestion
def get_unit_name(
    measurement_unit: str,
    length: Literal['short', 'long', 'narrow'] = 'long',
    locale: Locale | str | None = LC_NUMERIC,
) -> str | None:
```

File: babel/units.py [hunk 33]
```diff
@@ -62,7 +70,9 @@ def _find_unit_pattern(unit_id, locale=LC_NUMERIC):
             return unit_pattern
 
 
-def format_unit(value, measurement_unit, length='long', format=None, locale=LC_NUMERIC):
+def format_unit(value: float | decimal.Decimal, measurement_unit: str,
+                length: Literal['short', 'long', 'narrow'] = 'long', format: str | None = None,
+                locale: Locale | str | None = LC_NUMERIC) -> str:
```
Comment (akx) [inline babel/units.py:75]:
```suggestion
def format_unit(
    value: float | decimal.Decimal,
    measurement_unit: str,
    length: Literal['short', 'long', 'narrow'] = 'long',
    format: str | None = None,
    locale: Locale | str | None = LC_NUMERIC,
) -> str:
```

File: babel/units.py [hunk 34]
```diff
@@ -132,7 +142,7 @@ def format_unit(value, measurement_unit, length='long', format=None, locale=LC_N
     return f"{formatted_value} {fallback_name or measurement_unit}"  # pragma: no cover
 
 
-def _find_compound_unit(numerator_unit, denominator_unit, locale=LC_NUMERIC):
+def _find_compound_unit(numerator_unit: str, denominator_unit: str, locale: Locale | str | None = LC_NUMERIC) -> str | None:
```
Comment (akx) [inline babel/units.py:145]:
```suggestion
def _find_compound_unit(
    numerator_unit: str,
    denominator_unit: str,
    locale: Locale | str | None = LC_NUMERIC,
) -> str | None:
```

File: babel/units.py [hunk 35]
```diff
@@ -181,10 +191,10 @@ def _find_compound_unit(numerator_unit, denominator_unit, locale=LC_NUMERIC):
...
 
 def format_compound_unit(
-    numerator_value, numerator_unit=None,
-    denominator_value=1, denominator_unit=None,
-    length='long', format=None, locale=LC_NUMERIC
-):
+    numerator_value: float | decimal.Decimal, numerator_unit: str | None = None,
+    denominator_value: float | decimal.Decimal = 1, denominator_unit: str | None = None,
+    length: Literal['short', 'long', 'narrow'] = 'long', format: str | None = None,
+    locale: Locale | str | None = LC_NUMERIC) -> str | None:
```
Comment (akx) [inline babel/units.py:197]:
```suggestion
    numerator_value: float | decimal.Decimal,
    numerator_unit: str | None = None,
    denominator_value: float | decimal.Decimal = 1,
    denominator_unit: str | None = None,
    length: Literal["short", "long", "narrow"] = "long",
    format: str | None = None,
    locale: Locale | str | None = LC_NUMERIC,
) -> str | None:
```

File: babel/util.py [hunk 36]
```diff
@@ -224,25 +230,25 @@ def wraptext(text, width=70, initial_indent='', subsequent_indent=''):
 class FixedOffsetTimezone(tzinfo):
     """Fixed offset in minutes east from UTC."""
 
-    def __init__(self, offset, name=None):
+    def __init__(self, offset: float, name: str | None = None):
```
Comment (akx) [inline babel/util.py:233]:
```suggestion
    def __init__(self, offset: float, name: str | None = None) -> None:
    
```

**Review by akx — CHANGES_REQUESTED** [review]
Thanks for putting this work in!

I made a bunch of suggestions around readability (you can probably use GitHub's "commit suggestions in batch" thing to apply them), and then there were some somewhat suspect types wrt. docstrings.

File: babel/messages/checkers.py [hunk 37]
```diff
@@ -57,7 +60,7 @@ def python_format(catalog, message):
             _validate_format(msgid, msgstr)
 
 
-def _validate_format(format, alternative):
+def _validate_format(format, alternative) -> None:
```
[the remark this answers is not in the document]
Comment (DenverCoder1) [reply babel/messages/checkers.py:63]: Some functions not documented / part of the public API were skipped, but I can look into getting a few more in

File: babel/messages/mofile.py (hunk 4 above)
Comment (DenverCoder1) [reply babel/messages/mofile.py:19]: I believe it's a dependency of type-checkers, so as long as `__future__.annotations` is used and it's inside `if TYPE_CHECKING`, it should be fine.

Docs on it are here: https://github.com/python/typeshed/tree/main/stdlib/_typeshed

SupportsItems, SupportsWrite, SupportsRead, and SupportsReadline are marked as stable

File: babel/messages/pofile.py (hunk 6 above)
Comment (DenverCoder1) [reply babel/messages/pofile.py]: What type do you suggest? `IO[bytes] | IO[str]` seems to work.

**Comment by DenverCoder1** [comment]
> I made a bunch of suggestions around readability

This is a good start, maybe we can consider running Black on the whole project to give it some consistency?

File: babel/messages/pofile.py (hunk 6 above)
Comment (akx) [reply babel/messages/pofile.py]: I think [`IO[AnyStr]` is the usual idiom](https://docs.python.org/3.11/library/typing.html#typing.IO)?

**Review by akx — CHANGES_REQUESTED** [review]
@DenverCoder1 Let's ship this! ... Can you merge/rebase once more with #943 in?

**Review by akx — APPROVED** [review]
🚀

**Pull request merged**